x-pack/legacy/plugins/spaces/public/lib/copy_saved_objects_to_space/summarize_copy_result.ts

```diff
@@ -0,0 +1,142 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { ProcessedImportResponse, SavedObjectRecord } from 'ui/management/saved_objects_management';
+
+export interface SummarizedSavedObjectResult {
+  type: string;
+  id: string;
+  name: string;
+  conflicts: ProcessedImportResponse['failedImports'];
+  missingReferences: ProcessedImportResponse['failedImports'];
+  hasUnresolvableErrors: boolean;
+}
+
+interface SuccessfulResponse {
+  successful: true;
+  hasConflicts: false;
+  hasUnresolvableErrors: false;
+  objects: SummarizedSavedObjectResult[];
+  processing: false;
+}
+interface UnsuccessfulResponse {
+  successful: false;
+  hasConflicts: boolean;
+  hasUnresolvableErrors: boolean;
+  objects: SummarizedSavedObjectResult[];
+  processing: false;
+}
+
+interface ProcessingResponse {
+  objects: SummarizedSavedObjectResult[];
+  processing: true;
+}
+
+export type SummarizedCopyToSpaceResult =
+  | SuccessfulResponse
+  | UnsuccessfulResponse
+  | ProcessingResponse;
+
+export function summarizeCopyResult(
+  savedObject: SavedObjectRecord,
+  copyResult: ProcessedImportResponse | undefined,
+  includeRelated: boolean
+): SummarizedCopyToSpaceResult | null {
+  const successful = Boolean(copyResult && copyResult.failedImports.length === 0);
+
+  const conflicts = copyResult
+    ? copyResult.failedImports.filter(failed => failed.error.type === 'conflict')
+    : [];
+
+  const missingReferences = copyResult
+    ? copyResult.failedImports.filter(failed => failed.error.type === 'missing_references')
+    : [];
+
+  const unresolvableErrors = copyResult
+    ? copyResult.failedImports.filter(failed => failed.error.type !== 'conflict')
+    : [];
+
+  const hasConflicts = conflicts.length > 0;
+
+  const hasUnresolvableErrors = Boolean(
+    copyResult && copyResult.failedImports.some(failed => failed.error.type !== 'conflict')
+  );
+
+  const objectMap = new Map();
+  objectMap.set(`${savedObject.type}:${savedObject.id}`, {
+    type: savedObject.type,
+    id: savedObject.id,
+    name: savedObject.meta.title,
+    conflicts: conflicts.filter(
+      c => c.obj.type === savedObject.type && c.obj.id === savedObject.id
+    ),
+    missingReferences,
```

Comment:
Currently, if you try to use `export` or `copy to space` where a reference is missing, we get a `400` Bad Request is thrown. So, it's hard to tell how the missing references will actually be returned once this is fixed.

Do we really want all missing reference conflicts to show up here? Wouldn't this be all missing references for the entire "object graph" including transitive references? 

For the direct references and transitive references, we end up listing all missing reference errors that are for the object itself. If that's correct, shouldn't the "missing reference" for the saved object we're trying to copy always be `null`/`undefined`?

Reply:
Hmm, I think we're talking about different scenarios here. Any `copy to space` code that mentions "missing references" is talking about conditions which produce this **import** error: https://github.com/elastic/kibana/blob/4996c5d62e745e07b746c12186d65195be270c84/src/core/server/saved_objects/import/types.ts#L68-L78

Specifically, there are two conditions which cause this (that I'm aware of), both of which are produced by [`validate_references`](https://github.com/elastic/kibana/blob/e137477b5a00b8c83e63acda1121c1d2ee10541c/src/core/server/saved_objects/import/validate_references.ts#L79):
- An object being imported directly depends on an index pattern which isn't provided in the upload/copy request, nor in the target space (e.g., unchecking "include references" when copying a visualization which references an index pattern into a fresh space)
- An object being imported directly depends on a saved search which isn't provided in the upload/copy request, nor in the target space (e.g., unchecking "include references" when copying an index pattern which references a saved search into a fresh space)

The normal import/export flow shows a screen which allows users to choose a new index pattern or saved search to associate, but this is one of the features that differs between `import/export` and `copy to space`.

I have this "missing references" code in place for the sole purpose of [showing a more informative tooltip](https://github.com/elastic/kibana/blob/139d606215138e5f076248d0d29e9d6df8405e8b/x-pack/legacy/plugins/spaces/public/views/management/components/copy_saved_objects_to_space/copy_status_indicator.tsx#L67) when reviewing the "copy to space" results.


> Wouldn't this be all missing references for the entire "object graph" including transitive references?

No, currently, the import/export service only deals with direct references to index patterns and saved searches. No other reference types are checked for this condition on import, and we don't fail if this happens on a transitive reference either. This feels inconsistent to me, but also outside the scope of the "Copy to Space" feature IMO.